### PR TITLE
Generate server-specific timeout codes

### PR DESCRIPTION
**Pull request by heinrich5991**
This way, servers can no longer hijack the timeout codes of clients.
Timeout codes are generated from md5(seed + dummy + server address).

If `cl_timeout_seed` is set to the empty string, the old config options
are used instead.

### CODE REVIEW

File: src/engine/client/client.cpp [hunk 1]
```diff
@@ -678,8 +682,54 @@ void CClient::EnterGame()
...
+
+			char aBuf[64];
+			str_format(aBuf, sizeof(aBuf), "timeout code '%s' (%s)", m_aTimeoutCodes[i], i == 0 ? "normal" : "dummy");
+			m_pConsole->Print(IConsole::OUTPUT_LEVEL_ADDINFO, "client", aBuf);
+		}
+	}
+	else
+	{
+		str_copy(m_aTimeoutCodes[0], g_Config.m_ClTimeoutCode, sizeof(m_aTimeoutCodes[0]));
+		str_copy(m_aTimeoutCodes[0], g_Config.m_ClDummyTimeoutCode, sizeof(m_aTimeoutCodes[0]));
```
Comment (east) [inline src/engine/client/client.cpp]: `m_aTimeoutCodes[0]` is written to twice here

File: src/base/system.c [hunk 2]
```diff
@@ -2555,7 +2555,39 @@ int secure_random_init()
...
+void generate_password(char *buffer, unsigned length, unsigned short *random, unsigned random_length)
+{
+	static const char VALUES[] = "ABCDEFGHKLMNPRSTUVWXYZabcdefghjkmnopqt23456789";
+	static const size_t NUM_VALUES = sizeof(VALUES) - 1; // Disregard the '\0'.
+	dbg_assert(length >= random_length * 2 + 1, "too small buffer");
+	dbg_assert(NUM_VALUES * NUM_VALUES >= 2048, "need at least 2048 possibilities for 2-character sequences");
+
+	buffer[random_length * 2] = 0;
+
+	for(size_t i = 0; i < random_length; i++)
```
Comment (east) [inline src/base/system.c]: Error: `src/base/system.c:2567:2: error: ‘for’ loop initial declarations are only allowed in C99 or C11 mode
  for(size_t i = 0; i < random_length; i++)
  ^`

**Comment by east** [comment]
Regarding the user-interface, we currently have two edit-boxes for editing the main and dummy timeout codes. These are useless now, as long as `cl_timeout_seed` is set.

What should we do with them? It would be complicated if we had both configurations for timeout-code and timeout-code seed, so I propose to remove them and simply add a button for generating a new seed (the seed doesn't need to be showed to the user).

**Comment by east** [comment]
Looks good to me. Anyone else who wants to participate?

**Comment by BannZay** [comment]
Is still possible to reconnect if client was reloaded? (Client crash is real)

**Comment by east** [comment]
Yes, the generated timeout-codes will only change once you generate a new seed.

**Pull request merged**